app/code/Magento/MediaStorage/Model/File/Validator/Image.php

```diff
@@ -60,10 +60,16 @@ public function isValid($filePath): bool
 
         if (in_array($fileMimeType, $this->imageMimeTypes)) {
             try {
-                //phpcs:ignore Magento2.Functions.DiscouragedFunction
-                $image = imagecreatefromstring($this->file->fileGetContents($filePath));
+                if ($fileMimeType !== 'image/vnd.microsoft.icon') {
+                    //phpcs:ignore Magento2.Functions.DiscouragedFunction
+                    $image = imagecreatefromstring($this->file->fileGetContents($filePath));
+                } else {
+                    $imagick = new \Imagick();
+                    $imagick->setFormat('ICO');
+                    $image = $imagick->readImageBlob($this->file->fileGetContents($filePath));
```

Comment:
I don't think it's a good option because the Imagick extension is currently optional, and by default, Magento uses lib gd.
Can it handle this case?

Reply:
Hi @ihor-sviziev 
Lib **GD** does not support the `ICO` format and error returns.
After a long search for a solution, I think **Imagick** is the most suitable option, or we will have to install a third-party extension.

Reply:
@mmezhensky I'm OK with using imagick when it's available. But when it's not?

Reply:
@ihor-sviziev `Imagick` is already used in Magento.

Reply:
Yes, but only as **optional** dependence

Reply:
@mmezhensky I'm agree with @ihor-sviziev and believe that we could use `Magento\Framework\Image\Factory` here because it will use default adapter selected in configs.
If call of `open` throw exception than image is not valid.

Reply:
Hi @Den4ik 
Thanks for your information, but I tried to do it, and I always get an error for `ICO` files because **`GD2` does not support this format**, and `Magento\Framework\Image\Adapter\ImageMagick` has no function for setting `ICO` format.

To use `Magento\Framework\Image\Factory`, we always need to pass `$adapterName = "IMAGEMAGICK"` and extend `Magento\Framework\Image\Adapter\ImageMagick` class to add the ability to install the form.
But I think it will be a significant change in the Magento Framework that may cause problems in the future, so I suggest using the fix option I suggested earlier.

СС @ihor-sviziev

Reply:
Hi @mmezhensky. Thanks for your work.

> we always need to pass $adapterName = "IMAGEMAGICK"

It's not quite true. AdapterFactory retrieve current selected adapter from stores configuration. And my mind is that we receive exception if imagick is not selected as adapter at Stores config. It's guarant that user/administrator that select imagick as adapter will be sure that it's installed for php.

Reply:
@Den4ik 
Yes, but even if imagick is installed in the configuration and it is installed on the server, we will still get an exception because we need to set ICO format for the imagick object (but we can't do it without changes in `Magento\Framework\Image\Adapter\ImageMagick`).

Reply:
@mmezhensky I agree that it require some changes for `ImageMagick` class, but I don't see any problems use something like that in `open` method

```
            if (is_callable('exif_imagetype')) {
                $fileType = exif_imagetype($this->_fileName);
                if ($fileType === IMAGETYPE_ICO) {
                    $filename = 'ico:' . $filename;
                }
            }
            $this->_imageHandler = new \Imagick($filename);
```
It provide total support of ico files for image magick adapter
@ihor-sviziev @sidolov What do you think?

Reply:
I prepared these changes.
Thx!